pkg/rancher-desktop/backend/containerClient/registry.ts

```diff
@@ -0,0 +1,97 @@
+import registryAuth from '@pkg/backend/containerClient/auth';
+import { parseImageReference } from '@pkg/utils/dockerUtils';
+import fetch, { Headers } from '@pkg/utils/fetch';
+
+/**
+ * Registry interaction, with both Docker Hub and Docker Registry V2 APIs.
+ */
+class DockerRegistry {
+  /**
+   * Fetch some API endpoint from the registry
+   * @param endpoint The API endpoint, including the registry host.
+   */
+  async get(endpoint: URL): ReturnType<typeof fetch> {
+    const headers = await this.authenticate(endpoint);
+
+    return await fetch(endpoint.toString(), { headers });
+  }
+
+  /**
+   * List all tags for the given image name.
+   * @param name An image name, including registry as needed.
+   */
+  async getTags(name: string): Promise<string[]> {
+    const info = parseImageReference(name);
+    const tags: string[] = [];
+
+    if (!info) {
+      throw new Error(`Invalid image name: "${ name }"`);
+    }
+
+    let endpoint = new URL(`/v2/${ info.name }/tags/list?n=65536`, info.registry);
+    let hasMore = true;
+
+    while (hasMore) {
+      const resp = await this.get(endpoint);
+
+      if (!resp.ok) {
+        throw new Error(`Failed to fetch ${ endpoint }: ${ resp.status } ${ resp.statusText }`);
+      }
+
+      const result: { name: string, tags: string[] } = await resp.json();
+
+      if (result.name !== info.name) {
+        throw new Error(`Invalid tags: incorrect response name ${ result.name } from ${ endpoint }`);
+      }
+
+      tags.push(...result.tags);
+      hasMore = false;
+
+      for (const link of resp.headers[getAsList]('Link') ?? []) {
+        const fields = link.split(/;\s*/);
+
+        if (!fields.includes('rel="next"')) {
+          continue;
+        }
+        endpoint = new URL(fields[0].replace(/^<(.+)>$/, '$1'), endpoint);
+        hasMore = true;
+      }
+    }
+
+    return tags;
+  }
+
+  protected authenticate(endpoint: URL): Promise<Headers> {
+    return registryAuth.authenticate(endpoint);
+  }
+}
+
+const registry = new DockerRegistry();
+
+export default registry;
+
+// Extend Headers with a helper to get the header values as a list.
```

Comment:
Is this all being done just to make it easier to find any `Link` fields in the returned headers?

So first, does it convert the headers into a hash of `HeaderName => list of string values`?  If so, maybe it should be moved into `utils/` as a useful function that we might to have available elsewhere. If not, what is it doing?

Reply:
`Headers` internally keeps the headers it sees as a `Record<string, string[]>`; however, all of its public APIs only return a string (joining any arrays with `, ` i.e. comma + space).  So this is just a helper to get the original array.